File: apps/api_py/src/linkbook/integrations/mocks/gmail.py
"""Gmail mock — mirrors apps/api/src/integrations/_mocks/gmail.ts."""

from __future__ import annotations

import base64
import re
import time
from datetime import datetime, timezone

from ..http import HttpRequest, HttpResponse
from .store import get_mock_store
from .transport import consume_failure_for, json_response


async def handle(req: HttpRequest, path: str) -> HttpResponse:
    fail = consume_failure_for(f"gmail:{req.method} {path}")
    if fail is not None:
        return fail
    store = get_mock_store()

    if req.method == "POST" and path.endswith("/messages/send"):
        body = req.body or {}
        new_id = f"gm_{int(time.time() * 1000)}"
        store.sent_emails.append(
            {
                "to": body.get("to"),
                "cc": body.get("cc", []),
                "subject": body.get("subject", ""),
                "body": body.get("body", ""),
                "thread_id": body.get("thread_id"),
                "at": datetime.now(timezone.utc).isoformat(),
            }
        )
        return json_response(
            200,
            {
                "id": new_id,
                "threadId": body.get("thread_id") or f"t_{new_id}",
                "labelIds": ["SENT"],
            },
        )

    m = re.match(r"^/gmail/v1/users/me/threads/([^/]+)/modify$", path)
    if req.method == "POST" and m:
        return json_response(200, {"id": m.group(1), "labelIds": ["INBOX"]})

    m = re.match(r"^/gmail/v1/users/me/threads/([^/]+)$", path)
    if req.method == "GET" and m:
        thread_id = m.group(1)
        msgs = [m for m in store.gmail_messages.values() if m.thread_id == thread_id]
        return json_response(
            200,
            {
                "id": thread_id,
                "messages": [
                    {
                        "id": msg.id,
                        "threadId": msg.thread_id,
                        "labelIds": msg.label_ids,
                        "payload": {
                            "headers": [
                                {"name": "From", "value": msg.from_},
                                {"name": "To", "value": ", ".join(msg.to)},
                                {"name": "Cc", "value": ", ".join(msg.cc)},
                                {"name": "Subject", "value": msg.subject},
                                {"name": "Date", "value": msg.sent_at},
                            ],
                            "body": {"data": base64.b64encode(msg.body.encode()).decode()},
                        },
                    }
                    for msg in msgs
                ],
            },
        )

    return json_response(404, {"error": f"gmail: unmocked {req.method} {path}"})

File: apps/api/src/integrations/_mocks/gmail.ts
import type { HttpRequest, HttpResponse } from '../_http/client.js';
import { jsonResponse, consumeFailureFor } from './transport.js';
import { getMockStore } from './store.js';

// Mock Gmail endpoints (subset of users.messages):
//   POST /gmail/v1/users/me/messages/send         — send a message
//   POST /gmail/v1/users/me/threads/{id}/modify   — apply labels
//   GET  /gmail/v1/users/me/threads/{id}          — fetch thread (bodies)

export async function handleGmail(req: HttpRequest, path: string): Promise<HttpResponse> {
  const fail = consumeFailureFor(`gmail:${req.method} ${path}`);
  if (fail) return fail;

  const store = getMockStore();

  if (req.method === 'POST' && path.endsWith('/messages/send')) {
    const body = req.body as {
      to: string;
      cc?: string[];
      subject: string;
      body: string;
      thread_id?: string | null;
    };
    const id = `gm_${Date.now()}`;
    const sent = {
      to: body.to,
      cc: body.cc ?? [],
      subject: body.subject,
      body: body.body,
      thread_id: body.thread_id ?? null,
      at: new Date().toISOString(),
    };
    store.sent_emails.push(sent);
    return jsonResponse(200, {
      id,
      threadId: body.thread_id ?? `t_${id}`,
      labelIds: ['SENT'],
    });
  }

  const modify = path.match(/^\/gmail\/v1\/users\/me\/threads\/([^/]+)\/modify$/);
  if (req.method === 'POST' && modify) {
    return jsonResponse(200, { id: modify[1], labelIds: ['INBOX'] });
  }

  const fetchThread = path.match(/^\/gmail\/v1\/users\/me\/threads\/([^/]+)$/);
  if (req.method === 'GET' && fetchThread) {
    const id = fetchThread[1] ?? '';
    const messages = Array.from(store.gmail_messages.values()).filter((m) => m.thread_id === id);
    return jsonResponse(200, {
      id,
      messages: messages.map((m) => ({
        id: m.id,
        threadId: m.thread_id,
        labelIds: m.label_ids,
        payload: {
          headers: [
            { name: 'From', value: m.from },
            { name: 'To', value: m.to.join(', ') },
            { name: 'Cc', value: m.cc.join(', ') },
            { name: 'Subject', value: m.subject },
            { name: 'Date', value: m.sent_at },
          ],
          body: { data: Buffer.from(m.body).toString('base64') },
        },
      })),
    });
  }

  return jsonResponse(404, { error: `gmail: unmocked ${req.method} ${path}` });
}
